sklearn/neighbors/tests/test_neighbors.py

```diff
@@ -2259,3 +2259,80 @@ def _weights(dist):
     est = KNeighborsRegressor(n_neighbors=1, algorithm="brute", weights=_weights)
     est.fit(X, y)
     assert_allclose(est.predict([[0, 2.5]]), [6])
+
+
+@pytest.mark.parametrize(
+    "Estimator, params",
+    [
+        (neighbors.KNeighborsClassifier, {"n_neighbors": 2}),
+        (neighbors.KNeighborsRegressor, {"n_neighbors": 2}),
+        (neighbors.RadiusNeighborsRegressor, {}),
+        (neighbors.RadiusNeighborsClassifier, {}),
+        (neighbors.NearestCentroid, {}),
+        (neighbors.KNeighborsTransformer, {"n_neighbors": 2}),
+        (neighbors.RadiusNeighborsTransformer, {"radius": 1.5}),
+        (neighbors.LocalOutlierFactor, {"n_neighbors": 1}),
+        (neighbors.NearestNeighbors, {"n_neighbors": 1}),
+    ],
+)
+def test_nan_euclidean_support(Estimator, params):
+    """Check that the different neighbor estimators are lenient towards `nan`
+    values if using `metric="nan_euclidean"`.
+    """
+
+    X = [[0, 1], [1, np.nan], [2, 3], [3, 5]]
+    y = [0, 0, 1, 1]
+
+    params.update({"metric": "nan_euclidean"})
+    estimator = Estimator().set_params(**params).fit(X, y)
+
+    for response_method in ("kneighbors", "predict", "transform", "fit_predict"):
+        if hasattr(estimator, response_method):
+            output = getattr(estimator, response_method)(X)
+            if hasattr(output, "toarray"):
+                output = output.toarray()
+            assert not np.isnan(output).any()
+
+
+@pytest.mark.parametrize(
+    "Estimator, params",
+    [
+        (neighbors.KNeighborsClassifier, {"n_neighbors": 2}),
+        (neighbors.KNeighborsRegressor, {"n_neighbors": 2}),
+        # (neighbors.RadiusNeighborsRegressor, {}),
+        # (neighbors.RadiusNeighborsClassifier, {}),
+        (neighbors.NearestCentroid, {}),
+        # (neighbors.KNeighborsTransformer, {"n_neighbors": 2}),
```

Comment:
Thanks for the new comments, @glemaitre. I tried to address them in my last commit.

Regarding the test for empty inputs, when I tested it the [first time](https://github.com/scikit-learn/scikit-learn/pull/25330#discussion_r1096493728), I must have done something wrong in the notebook and didn't realize that the behavior is a bit strange for some of the classes:

```python
from sklearn import __version__
__version__
>>> '1.3.dev0'

import numpy as np
from sklearn import neighbors

X = [[np.nan, np.nan], [np.nan, np.nan], [np.nan, np.nan], [np.nan, np.nan]]
y = [1, 2, 3, 4]

print(neighbors.RadiusNeighborsRegressor(metric='nan_euclidean').fit(X, y).predict(X))
>>> [-2147483648 -2147483648 -2147483648 -2147483648]
print(neighbors.RadiusNeighborsClassifier(metric="nan_euclidean").fit(X, y).predict(X))
>>> ValueError: No neighbors found for test samples array([0, 1, 2, 3], dtype=int64), you can try using larger radius, giving a label for outliers, or considering removing them from your dataset.

for n in range(1, 5):
    model = neighbors.KNeighborsTransformer(metric="nan_euclidean", n_neighbors=n)
    print(f"n_neighbors={n}")
    print(model.fit_transform(X).toarray())
>>> n_neighbors=1
>>> [[nan nan  0.  0.]
 [nan nan  0.  0.]
 [nan nan  0.  0.]
 [nan nan  0.  0.]]
>>> n_neighbors=2
>>> [[nan nan nan  0.]
 [nan nan nan  0.]
 [nan nan nan  0.]
 [nan nan nan  0.]]
>>> n_neighbors=3
>>> [[nan nan nan nan]
 [nan nan nan nan]
 [nan nan nan nan]
 [nan nan nan nan]]
>>> n_neighbors=4
>>> ValueError: Expected n_neighbors <= n_samples,  but n_samples = 4, n_neighbors = 5
# Note that n_samples = 4 and actually n_neighbors  should be = 4 also
```

It seems that `RadiusNeighborsRegressor` and `RadiusNeighborsClassifier` cannot find anyone with a distance less than the radius (and they behave differently from each other). As for `KNeighborsTransformer`, I don't know how to explain what's happening... :dizzy_face:

What should we do?

Reply:
Personally, I think that it would be fine to raise an error for this case but I don't know we can easily detect it.